Ticketing/Ticketing.BAL/Services/EmailService.cs

```diff
@@ -0,0 +1,108 @@
+    {
+        private readonly IOptions<EmailSettings> _optionsEmailSettings;
+        private readonly IOptions<RetryPolicySettings> _optionsRetryPolicySettings;
+        private readonly AsyncRetryPolicy<Response> _retryPolicy;
+        private readonly IServiceScopeFactory _serviceScopeFactory;
+        private readonly ILog _logger;
+
+        public EmailService(IOptions<EmailSettings> optionsEmailSettings,
+            IOptions<RetryPolicySettings> optionsRetryPolicySettings,
+            IServiceScopeFactory serviceScopeFactory,
+            ILog logger)
+        {
+            _optionsEmailSettings = optionsEmailSettings;
+            _optionsRetryPolicySettings = optionsRetryPolicySettings;
+            _serviceScopeFactory = serviceScopeFactory;
+            _logger = logger;
+
+            _retryPolicy = RetryPolicy(_optionsRetryPolicySettings.Value.RetryCount, _optionsRetryPolicySettings.Value.SleepDurationSec);
+        }
+
+        public async Task SendEmailAsync(EmailNotification emailNotification)
+        {
+            _logger.Info($"EmailService Start Send Email.");
+            EmailNotification savedEmailNotification = await SaveNotificationAsync(emailNotification);
+
+            string? fromEmail = _optionsEmailSettings.Value.SenderEmail;
+            string? fromName = _optionsEmailSettings.Value.SenderName;
+            string? apiKey = _optionsEmailSettings.Value.ApiKey;
+            var sendGridClient = new SendGridClient(apiKey);
+            var from = new EmailAddress(fromEmail, fromName);
+            var to = new EmailAddress(emailNotification.Email, emailNotification.Name);
+            var subject = "Email Notification";
+
+            var body = $"This is Email Notification order {emailNotification.OrderId} amount {emailNotification.OrderAmount} summary {emailNotification.OrderSummary}";
+            var msg = MailHelper.CreateSingleEmail(from, to, subject, body, "");
+
+
+            savedEmailNotification.EmailRequestStatus = EmailRequestStatus.InProgress;
+            await UpdateNotificationAsync(savedEmailNotification);
+
+            var response = await _retryPolicy.ExecuteAsync(async () => await sendGridClient.SendEmailAsync(msg));
+
+            if (response.IsSuccessStatusCode)
+            {
+                _logger.Info($"EmailService Success.");
+                savedEmailNotification.EmailRequestStatus = EmailRequestStatus.Success;
+                await UpdateNotificationAsync(savedEmailNotification);
+            }
+            else
+            {
+                _logger.Info($"EmailService Fail.");
+                savedEmailNotification.EmailRequestStatus = EmailRequestStatus.Fail;
+                await UpdateNotificationAsync(savedEmailNotification);
+            }
+        }
+
+        private async Task<EmailNotification> SaveNotificationAsync(EmailNotification emailNotification)
+        {
+            using (var scope = _serviceScopeFactory.CreateScope())
```

Comment:
???? why I see IOC here??? Injections and only injetions please

Reply:
Because have no other methods get Scoped service (Repository) for Singleton